app/properties/claim/actions.ts

```diff
@@ -0,0 +1,237 @@
+        landlord_id: z.string(),
+        started_at: z.date(),
+        ended_at: z.date().nullable()
+    })
+    // started_at must be in the past
+    .superRefine(({ started_at }, ctx) => {
+        if (started_at > new Date()) {
+            ctx.addIssue({
+                code: z.ZodIssueCode.custom,
+                message: 'Start date must be in the past',
+                path: ['started_at']
+            })
+        }
+    })
+    // If ended_at is not null, it must be in the past
+    .superRefine(({ ended_at }, ctx) => {
+        if (ended_at && ended_at > new Date()) {
+            ctx.addIssue({
+                code: z.ZodIssueCode.custom,
+                message: 'End date must be in the past',
+                path: ['ended_at']
+            })
+        }
+    })
+    // If ended_at is not null, it must be after started_at
+    .superRefine(({ started_at, ended_at }, ctx) => {
+        if (ended_at && started_at > ended_at) {
+            ctx.addIssue({
+                code: z.ZodIssueCode.custom,
+                message: 'End date must be after start date',
+                path: ['ended_at']
+            })
+        }
+    })
+
+export const claimProperty = async (
+    propertyId: string,
+    landlordId: string,
+    prevState: State,
+    formData: FormData
+): Promise<State> => {
+    const validatedFields = ClaimPropertySchema.safeParse({
+        property_id: propertyId,
+        landlord_id: landlordId,
+        started_at: formData.get('started_at'),
+        ended_at: formData.get('ended_at')
+    });
+
+    if (!validatedFields.success) return {
+        errors: validatedFields.error.flatten().fieldErrors,
+        message: 'Invalid Fields. Failed to Claim Property.',
+    }
+
+    let {
+        property_id,
+        landlord_id,
+        started_at,
+        ended_at
+    } = {
+        ...validatedFields.data,
+        started_at: validatedFields.data.started_at,
+        ended_at: validatedFields.data.ended_at || new Date()
+    }
+
+    if (!property_id) {
+        return {
+            errors: {
+                property_id: ['Property ID is required']
+            },
+            message: 'Invalid Fields Submitted'
+        }
+    }
+    // Need to check more things here before we can claim a property
+
+    const cookieStore = cookies();
+    const supabase = createServerClient(cookieStore);
+
+    // check if the user is logged in
+    const { data: { user }, error: userError } = await supabase.auth.getUser()
+    if (userError)
+        return {
+            message: 'Error Fetching User'
+        };
+
+
+    if (!user)
+        return {
+            message: 'User Not Logged In'
+        };
+
+    //  check a the user is already the landlord of the property for this time period
+    const { data: propertyOwnershipList, error: propertyOwnershipError } = await supabase
+        .from('property_ownership')
+        .select('landlord_id, started_at, ended_at')
+        .eq('property_id', property_id)
```

Comment:
there are currently no rls policies on this table to allow selection.

we either want to add them (you can remove or change the migration I added since we have not yet pushed it to development/production)

or you can use a service supabase client like I do in lines 181-193 to overwrite RLS

I'd recommend the first option since we want this table to be readable anyway, but we can do either for now